### PR TITLE
Option to disable checksum verification

### PR DESCRIPTION
**Pull request by HeyaGlitz**
This fixes a few issues that have been recently occurring, however this option shouldn't be used carelessly since it also indicates a corrupt/malware infected/tampered download.
Fixes "java.lang.RuntimeException: Unable to detect the forge installer!" and "it was impossible to fetch the required libraries ... something to do with forge".

### CODE REVIEW

File: launcher/ui/pages/global/APIPage.cpp [hunk 1]
```diff
@@ -183,6 +185,7 @@ void APIPage::applySettings()
     QString flameKey = ui->flameKey->text();
     s->set("FlameKeyOverride", flameKey);
     s->set("UserAgentOverride", ui->userAgentLineEdit->text());
+    s->set("DisableChecksumVerification", ui->disableChecksumVerificationCheckBox->isChecked());
```
Comment (slendidev) [inline launcher/ui/pages/global/APIPage.cpp]: Before setting, wouldn't it be *much* better to prompt the user with a warning
not to do this? This is a *really* dangerous thing.

File: launcher/net/Download.cpp [hunk 2]
```diff
@@ -85,6 +85,12 @@ auto Download::makeFile(QUrl url, QString path, Options options) -> Download::Pt
 
 void Download::addValidator(Validator* v)
 {
+    auto s = APPLICATION->settings();
+    if(s->get("DisableChecksumVerification").toBool())
+    {
+        qDebug() << "Checksum verification disabled, no validator added";
+        return;
+    }
```
Comment (slendidev) [inline launcher/net/Download.cpp]: I would put this message at a warning level.

File: launcher/net/ChecksumValidator.h [hunk 3]
```diff
@@ -64,8 +65,10 @@ class ChecksumValidator : public Validator {
 
     auto validate(QNetworkReply&) -> bool override
     {
-        if (m_expected.size() && m_expected != hash()) {
+        if (m_expected.size() && m_expected != hash())
+        {
             qWarning() << "Checksum mismatch, download is bad.";
+            qDebug() << "Expected" << m_expected.toHex() << "but got" << hash().toHex();
```
Comment (slendidev) [inline launcher/net/ChecksumValidator.h]: I would combine both of those, so the message would look something like this
instead: "Checksum mismatch, expected %1 but got %2.". Putting it on warning
level would be nice too.

**Review by slendidev — COMMENTED** [review]
Some general clean-up ideas.

File: launcher/ui/pages/global/APIPage.cpp [hunk 4]
```diff
@@ -185,7 +185,22 @@ void APIPage::applySettings()
...
     s->set("FlameKeyOverride", flameKey);
     s->set("UserAgentOverride", ui->userAgentLineEdit->text());
-    s->set("DisableChecksumVerification", ui->disableChecksumVerificationCheckBox->isChecked());
+    bool previous = s->get("DisableChecksumVerification").toBool();
+    if(ui->disableChecksumVerificationCheckBox->isChecked() && previous == false)
+    {
+        QMessageBox::StandardButton reply = QMessageBox::question(this
+                                                                  , "Warning"
+                                                                  , "Are you sure you want to disable checksum validation? \nTHIS IS VERY DANGEROUS, so do it at your own risk."
+                                                                  , QMessageBox::Yes|QMessageBox::No);
```
Comment (slendidev) [inline launcher/ui/pages/global/APIPage.cpp]: Maybe put the comma at the end of the line and not the beginning?

File: launcher/ui/pages/global/APIPage.cpp [hunk 5]
```diff
@@ -185,7 +185,22 @@ void APIPage::applySettings()
...
+        if(reply == QMessageBox::Yes)
+        {
+            s->set("DisableChecksumVerification", ui->disableChecksumVerificationCheckBox->isChecked());
+        }
+    }
+    else
+    {
+        s->set("DisableChecksumVerification", ui->disableChecksumVerificationCheckBox->isChecked());
+    }
```
Comment (slendidev) [inline launcher/ui/pages/global/APIPage.cpp:203]: Surely there is a better way to do this, seems like duplicate code to me.

Comment (HeyaGlitz) [reply launcher/ui/pages/global/APIPage.cpp:203]: Dont know how else one could do it

Comment (HeyaGlitz) [reply launcher/ui/pages/global/APIPage.cpp:203]: The only thing I could do is change 197 from `ui->disableChecksumVerificationCheckBox->isChecked()` to just `true`

**Comment by LennyMcLennington** [comment]
I thought about this for a long time and I'm still not sure whether this is a great idea, the checksums should never be wrong and if they are it would probably be an indication of a different problem that should be fixed somewhere else, rather than just forcing it by ignoring checksums... The root cause should be fixed instead, sometimes workarounds are good for a quick fix but in this sort of thing I just don't think it's worth it. Sorry.

**Pull request closed without merging**